endpoints/test_utils.go

```diff
@@ -0,0 +1,193 @@
+		t.Fatalf("Expected 400 response. Got: %d, Msg: %v", putTrace.Code, putTrace.Body.String())
+		return
+	}
+}
+
+func benchmarkPutHandler(b *testing.B, testCase string) {
+	b.StopTimer()
+	//Set up a request that should succeed
+	request, err := http.NewRequest("POST", "/cache", strings.NewReader(testCase))
+	if err != nil {
+		b.Errorf("Failed to create a POST request: %v", err)
+	}
+
+	//Set up server ready to run
+	router := httprouter.New()
+	backend := backends.NewMemoryBackend()
+
+	router.POST("/cache", NewPutHandler(backend, 10, true))
+	router.GET("/cache", NewGetHandler(backend, true))
+
+	rr := httptest.NewRecorder()
+
+	//for statement to execute handler function
+	for i := 0; i < b.N; i++ {
+		b.StartTimer()
+		router.ServeHTTP(rr, request)
+		b.StopTimer()
+	}
+}
+
+func newMockBackend() *backends.MemoryBackend {
+	backend := backends.NewMemoryBackend()
+
+	backend.Put(context.TODO(), "non-36-char-key-maps-to-json", `json{"field":"value"}`, 0)
+	backend.Put(context.TODO(), "36-char-key-maps-to-non-xml-nor-json", `#@!*{"desc":"data got malformed and is not prefixed with 'xml' nor 'json' substring"}`, 0)
+	backend.Put(context.TODO(), "36-char-key-maps-to-actual-xml-value", "xml<tag>xml data here</tag>", 0)
+
+	return backend
+}
+
+type faultyRequestBodyReader struct {
+	mock.Mock
+}
+
+func (b *faultyRequestBodyReader) Read(p []byte) (n int, err error) {
+	args := b.Called(p)
+	return args.Int(0), args.Error(1)
+}
+
+func (b *faultyRequestBodyReader) Close() error {
+	args := b.Called()
+	return args.Error(0)
+}
+
+type errorReturningBackend struct{}
+
+func (b *errorReturningBackend) Get(ctx context.Context, key string) (string, error) {
+	return "", fmt.Errorf("This is a mock backend that returns this error on Get() operation")
+}
+
+func (b *errorReturningBackend) Put(ctx context.Context, key string, value string, ttlSeconds int) error {
+	return fmt.Errorf("This is a mock backend that returns this error on Put() operation")
+}
+
+func NewErrorReturningBackend() *errorReturningBackend {
+	return &errorReturningBackend{}
+}
+
+type deadlineExceedingBackend struct{}
+
+func (b *deadlineExceedingBackend) Get(ctx context.Context, key string) (string, error) {
+	return "", nil
+}
+
+func (b *deadlineExceedingBackend) Put(ctx context.Context, key string, value string, ttlSeconds int) error {
+	var err error
+
+	d := time.Now().Add(50 * time.Millisecond)
+	sampleCtx, cancel := context.WithDeadline(context.Background(), d)
+
+	// Even though ctx will be expired, it is good practice to call its
+	// cancellation function in any case. Failure to do so may keep the
+	// context and its parent alive longer than necessary.
+	defer cancel()
+
+	select {
+	case <-time.After(1 * time.Second):
+		//err = fmt.Errorf("Some other error")
+		err = nil
+	case <-sampleCtx.Done():
+		err = sampleCtx.Err()
+	}
+	return err
+}
+
+func NewDeadlineExceededBackend() *deadlineExceedingBackend {
+	return &deadlineExceedingBackend{}
+}
```

Comment:
Similar comment from earlier. This is not considered a test file by Go's compiler and will be included in the production binary. Can you move these helpers to the _test.go file they're used in? Are they shared among multiple test files?

Reply:
😮 I didn't know that. Moved everything in `endpoints/test_utils.go` back to `_test.go`.